### PR TITLE
Add gzip (zlib) bindings.

### PR DESCRIPTION
**Pull request by kentonv**

### CODE REVIEW

**Comment by kentonv** [comment]
(No makefile support yet because it introduces a new dependency...)

File: c++/src/kj/compat/gzip-test.c++ [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  gzipOut.end().wait(io.waitScope);
+
+  MockInputStream rawInput(rawOutput.bytes, kj::maxValue);
+  GzipInputStream gzipIn(rawInput);
+  auto decompressed = gzipIn.readAllBytes().wait(io.waitScope);
+
+  KJ_ASSERT(decompressed.size() == bytes.size());
+  KJ_ASSERT(memcmp(bytes.begin(), decompressed.begin(), bytes.size()) == 0);
+}
+
```
Comment (harrishancock) [inline c++/src/kj/compat/gzip-test.c++]: Maybe add a test, or a TODO, for the array-of-arrays overload of `write()`?

File: c++/src/kj/compat/gzip.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+#ifndef KJ_COMPAT_GZIP_H_
+#define KJ_COMPAT_GZIP_H_
+
+#include <kj/async-io.h>
+#include <zlib.h>
+
+namespace kj {
+
+class GzipInputStream: public AsyncInputStream {
```
[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/compat/gzip.h]: "GzipAsync", but yes, I agree -- we may very well want to add non-async bindings at some point. Done.

File: c++/src/kj/compat/gzip.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+public:
+  GzipInputStream(AsyncInputStream& inner);
+  ~GzipInputStream() noexcept(false);
+
```
[the remark this answers is not in the document]
Comment (kentonv) [reply c++/src/kj/compat/gzip.h]: Indeed, I always forget to do this.

File: c++/src/kj/compat/gzip-test.c++ (hunk 1 above)
Comment (kentonv) [reply c++/src/kj/compat/gzip-test.c++]: added test

**Pull request merged**